_source/_posts/2022-06-16-oauth-java.adoc

```diff
@@ -0,0 +1,389 @@
+
+https://twitter.com/briandemers/status/1520104453155835905[A browser is required]. The browser doesn't need to be on the same device. There are basically three options if a user is involved:
+
+- Web app: handle a redirect
+- CLI, TV, etc: use the Device Grant (browser on another device), app uses REST
+- Native app: use a custom URL handler
+
+== Java's OAuth 2.0 support
+
+Java, the language, does not have native OAuth support. In fact, the JDK doesn't even contain APIs to write a web app. The basic building blocks for constructing web apps is provided by the https://jakarta.ee/specifications/servlet/5.0/[Servlet API and Jakarta EE].
+
+The Servlet specification has a https://jakarta.ee/specifications/servlet/5.0/jakarta-servlet-spec-5.0.html#security[security section], but there's no OAuth in it. The https://jakarta.ee/specifications/security/3.0/[Jakara Security 3.0] (still under development at the time of this writing) does mention that OpenID Connect and OAuth did _not_ make it into the specification. To me, it seems strange that the "enterprise edition" of Java doesn't have support for delegated authorization or federated identity. Hopefully we work as a community to fix this in the future.
+
+On the upside, Java 18 does contain a `jwebserver` you can use to https://inside.java/2021/12/06/working-with-the-simple-web-server/[serve up static files]. That's about all the "web" you'll find in the JDK though.
+
+=== OAuth in Java frameworks
+
+When SAML was first invented in 2005, Java web frameworks were experiencing their peak of popularity. There were _hundreds_ of Java web frameworks and developers would make fun of this embarrassment of riches like they did with JavaScript frameworks in the late 2000s.
+
+image::{% asset_path 'blog/java-oauth2-2022/history-of-web-frameworks-timeline.png' %}[alt=History of web frameworks timeline,link={% asset_path 'blog/java-oauth2-2022/history-of-web-frameworks-timeline.png' %}]
+
+[role="text-right small"]
+Image https://github.com/mraible/history-of-web-frameworks-timeline[maintained on GitHub] by https://twitter.com/mraible[@mraible]
+
+Most Java web frameworks supported a form of MVC and others were component-based. All of them rendered their UIs on the server-side.
+
+Fast-forward to 2022 and Java is mostly used for APIs, while JavaScript is often used for UIs.
+
+"I do find it ironic that there are now several JavaScript frameworks that provide developers joy with server-side rendering. It's like the Java MVC frameworks of old, but new to JavaScript. History repeats itself, eh?"
+-- Matt Raible
+
+One of OAuth's key patterns is a *resource server*. A resource server accepts an access token. If it's valid, it gives a client access to the resource owner's data. In this example, a client is an app, the resource owner is a user, and the resource server is the Java API you develop.
+
+=== MicroProfile JWT authentication
```

Comment:
maybe a few lines on why you are comparing these frameworks specifically might be nice. Coz one might wonder that Spring is kind of defacto for building java web apps, but the other frameworks are fancy and new but are they actually used a lot, whats their market share?, should an average java developer care about all of them?

Reply:
Added this as a paragraph before the "MicroProfile JWT authentication" header.

> In the section below, I'll show you how you can use OAuth and set up a resource server with the most popular Java frameworks. You might think that Spring Boot is the de facto standard, and you should just use it. However, there's a lot of innovation happening elsewhere, and I think it's good to be aware of other options.